lib/gringotts/gateways/we_pay.ex

```diff
@@ -0,0 +1,426 @@
+  API method.
+
+  [docs]: hhttps://developer.wepay.com/
+  [req-resp]: https://developer.wepay.com/api/reference/structures
+
+  ## Supported currencies and countries
+
+  WePay supports the countries listed [here][all-country-list]
+
+  [all-country-list]: [https://support.wepay.com/hc/en-us/articles/203611643-Is-WePay-International-]
+
+  ## Following the examples
+
+  1. First, set up a sample application and configure it to work with WePay.
+  - You could do that from scratch by following our [Getting Started][gs] guide.
+      - To save you time, we recommend [cloning our example
+      repo][example] that gives you a pre-configured sample app ready-to-go.
+          + You could use the same config or update it the with your "secrets"
+          as described [above](#module-registering-your-monei-account-at-wepay).
+
+  2. Run an `iex` session with `iex -S mix` and add some variable bindings and
+  aliases to it (to save some time):
+  ```
+  iex> alias Gringotts.{Response, CreditCard, Gateways.WePay}
+  iex> card = %CreditCard{first_name: "Jo",
+                          last_name: "Doe",
+                          number: "4200000000000000",
+                          year: 2099, month: 12,
+                          verification_code: "123", brand: "VISA"}
+  ```
+  We'll be using these in the examples below.
+
+  [gs]: https://github.com/aviabird/gringotts/wiki/
+  [home]: https://go.wepay.com
+  [example]: https://github.com/aviabird/gringotts_example
+
+  """
+
+  # The Base module has the (abstract) public API, and some utility
+  # implementations.
+  use Gringotts.Gateways.Base
+
+  # The Adapter module provides the `validate_config/1`
+  # Add the keys that must be present in the Application config in the
+  # `required_config` list
+  use Gringotts.Adapter, required_config: [:access_token]
+
+  import Poison, only: [decode: 1]
+
+  alias Gringotts.{Money, CreditCard, Response}
+
+  @test_url "https://stage.wepayapi.com/v2"
+
+  @doc """
+  Performs a (pre) Authorize operation.
+
+  The authorization validates the `card` details with the banking network,
+  places a hold on the transaction `amount` in the customer’s issuing bank.
+
+  WePay returns an ID string which can be used to:
+
+  * `capture/3` _an_ amount.
+
+  ## Example
+  ```
+  iex> amount = Money.new(42, :USD)
+  iex> {:ok, auth_result} = Gringotts.authorize(Gringotts.Gateways.WePay, amount, card, opts)
+  iex> auth_result.id # This is the authorization ID
+  ```
+  """
+
+  @spec authorize(Money.t(), CreditCard.t(), keyword) :: {:ok | :error, Response}
+  def authorize(amount, card = %CreditCard{}, opts) do
+    {currency, value, _} = Money.to_integer(amount)
+
+    with {:ok, card_token_response} <- store(card, opts),
+         {:ok, card_token} <- extract_card_token(card_token_response) do
+      body =
+        Poison.encode!(%{
+          account_id: opts[:account_id],
+          short_description: opts[:short_description],
```

Comment:
1. This map is so verbose! Most of the keys are not transformed. Please consider the following style to build this map:
    ```elixir
    opts
    |> Keyword.take(@required_or_allowed_keys)
    |> Enum.into(%{})
    |> Map.merge(%{
      currency: currency,
      value: value,
      payment_method: %{...},
      ...
    })
    ```